aiida/common/progress_reporter.py

```diff
@@ -0,0 +1,111 @@
+# -*- coding: utf-8 -*-
+###########################################################################
+# Copyright (c), The AiiDA team. All rights reserved.                     #
+# This file is part of the AiiDA code.                                    #
+#                                                                         #
+# The code is hosted on GitHub at https://github.com/aiidateam/aiida-core #
+# For further information on the license, see the LICENSE.txt file        #
+# For further information please visit http://www.aiida.net               #
+###########################################################################
+# pylint: disable=global-statement,unused-argument
+"""Provide a singleton progress reporter implementation.
```

Comment:
Do you have a specific use case in mind that is not a progress bar?

If not, instead of "progress reporter" I would prefer the term "progress bar", which everybody immediately understands rather than trying to be more generic for a use case that we may never have.

A related question is: do you think there is a need to be generic with respect to the progress bar implementation, rather than hardcoding `tqdm`?

I certainly get the point of making the progress bar available not just in the import/export but in `aiida.common` but I'm wondering a bit whether all this boilerplate code is needed.
In the end, it might be a case of [YAGNI](https://en.wikipedia.org/wiki/You_aren%27t_gonna_need_it) ;-)

Reply:
Of the top of my head, maybe you want to report progress via a web based application (like aiidalab) or in a GUI. These would not use tqdm, and there is also nothing in the interface that requires it to be a bar.

I'm not that bothered about changing to bar, but also I really don't see it as that big an issue. I'm sure people can grasp the idea of a progress reporter.

I am absolutely definitely against hard coding tqdm. For starters, the CLI already requires that it be a partial tqdm.

The major problem with the import/export (and moving to a new format) is that half the stuff is hard-coded. I would much rather have YAGNI (within reason), if it means you don't have to later refactor everything.
Also, its hardly that much boilerplate, the file has less code than the original progress_bar.py one.

Reply:
Also talking about YAGNI in aiida-core, where we keep around two database backends (django/sqla) for really no reason now (since JSONB was implemented in django), is a bit ironic 😜

Reply:
> Also talking about YAGNI in aiida-core, where we keep around two database backends (django/sqla) for really no reason now (since JSONB was implemented in django), is a bit ironic 😜

Well, you know that I am very much in favor of taking a very hard look at dropping one of the two - to be fair, though, there was a clear use case when sqla was introduced.

As for the progress reporter/bar, I don't want to be the one to impose my opinion and since we want to get this merged, I'm ok with keeping it.
It is true that the code is relatively compact; it's just perhaps not very obvious how to use it (will add another comment in the final review)